Zend/zend_compile.c

```diff
@@ -4712,6 +4712,171 @@ static void zend_compile_ns_call(znode *result, znode *name_node, zend_ast *args
+	if (string_placeholder_count != (args->children - 1)) {
+		return FAILURE;
+	}
+
+	znode *elements = NULL;
+
+	if (string_placeholder_count > 0) {
+		elements = safe_emalloc(sizeof(*elements), string_placeholder_count, 0);
+	}
+
+	/* Compile the value expressions first for error handling that is consistent
+	 * with a function call: Values that fail to convert to a string may emit errors.
+	 */
+	for (uint32_t i = 0; i < string_placeholder_count; i++) {
+		zend_compile_expr(elements + i, args->child[1 + i]);
+		if (elements[i].op_type == IS_CONST) {
+			if (Z_TYPE(elements[i].u.constant) != IS_ARRAY) {
+				convert_to_string(&elements[i].u.constant);
+			}
+		}
+	}
+
+	uint32_t rope_elements = 0;
+	uint32_t rope_init_lineno = -1;
+	zend_op *opline = NULL;
+
+	string_placeholder_count = 0;
+	p = Z_STRVAL_P(format_string);
+	end = p + Z_STRLEN_P(format_string);
+	char *offset = p;
+	for (;;) {
+		p = memchr(p, '%', end - p);
+		if (!p) {
+			break;
+		}
+
+		char *q = p + 1;
+		ZEND_ASSERT(q < end);
+		ZEND_ASSERT(*q == 's' || *q == '%');
+
+		if (*q == '%') {
+			/* Optimization to not create a dedicated rope element for the literal '%':
+			 * Include the first '%' within the "constant" part instead of dropping the
+			 * full placeholder.
+			 */
+			p++;
+		}
+
+		if (p != offset) {
+			znode const_node;
+			const_node.op_type = IS_CONST;
+			ZVAL_STRINGL(&const_node.u.constant, offset, p - offset);
+			if (rope_elements == 0) {
+				rope_init_lineno = get_next_op_number();
+			}
+			opline = zend_compile_rope_add(result, rope_elements++, &const_node);
+		}
+
+		if (*q == 's') {
+			/* Perform the cast of constant arrays when actually evaluating corresponding placeholder
+			 * for correct error reporting.
+			 */
+			if (elements[string_placeholder_count].op_type == IS_CONST) {
+				if (Z_TYPE(elements[string_placeholder_count].u.constant) == IS_ARRAY) {
+					zend_emit_op_tmp(&elements[string_placeholder_count], ZEND_CAST, &elements[string_placeholder_count], NULL)->extended_value = IS_STRING;
+				}
+			}
+			if (rope_elements == 0) {
+				rope_init_lineno = get_next_op_number();
+			}
+			opline = zend_compile_rope_add(result, rope_elements++, &elements[string_placeholder_count]);
+
+			string_placeholder_count++;
+		}
+
+		p = q;
+		p++;
+		offset = p;
+	}
+	if (end != offset) {
+		/* Add the constant part after the last placeholder. */
+		znode const_node;
+		const_node.op_type = IS_CONST;
+		ZVAL_STRINGL(&const_node.u.constant, offset, end - offset);
+		if (rope_elements == 0) {
+			rope_init_lineno = get_next_op_number();
+		}
+		opline = zend_compile_rope_add(result, rope_elements++, &const_node);
+	}
+	if (rope_elements == 0) {
+		/* Handle empty format strings. */
+		znode const_node;
+		const_node.op_type = IS_CONST;
+		ZVAL_EMPTY_STRING(&const_node.u.constant);
+		if (rope_elements == 0) {
+			rope_init_lineno = get_next_op_number();
+		}
+		opline = zend_compile_rope_add(result, rope_elements++, &const_node);
```

Comment:
You can just assign the constant to `result` and dodge `zend_compile_rope_finalize()`. You can probably also move this way up and return early (e.g. to the length check).

Reply:
I opted to move this below the argument count check to uphold the safety check that the `sprintf()` call will be emitted if the number of placeholders does not match the number of arguments: This ensures that arguments will be consistently and reliably be evaluated.